File: src/onscala/codility/primencomposite/MinPerimeterRectangle.scala
package onscala.codility.primencomposite

import org.scalatest.Assertions
import org.testng.Assert.assertEquals
import org.testng.annotations.{DataProvider, Test}

/**
  * Project: scala_challenges
  * Package: onscala.codility.primencomposite
  *
  * User: luisvillaca
  * Date: 25/06/20
  * Time: 23:26
  */
class MinPerimeterRectangle extends Assertions {

  def solution(n: Int): Int = {
    val sqrtOfN = math.sqrt(n)
    // check if it's a factor - if so, calculate perimeter and fetch the minimum value
    (1 to sqrtOfN.toInt).filter(n % _ == 0)
      .map( factor => 2.toLong * (factor + (n/factor))).min.toInt
    // cast (toLong) to avoid possible overflow
    // cast back to int to respond
  }

  //Evaluating
  @DataProvider(name= "tests")
  def testData() = {
    Array(
      Array[Any](30, 22)
    )
  }

  @Test(dataProvider = "tests")
  def verify(intInput: Int, expectedOutput: Int): Unit = {
    val actualOutput = solution(intInput)
    assertEquals(actualOutput, expectedOutput)
  }

}

File: src/onjava/codility/primencomposite/MinPerimeterRectangle.java
package onjava.codility.primencomposite;

import org.testng.Assert;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

/**
 * Site: Codility
 * Problem: MinPerimeterRectangle
 *
 * User: Luis Villaca
 * Date: 01/05/2020
 */
public class MinPerimeterRectangle {
    public static int solution(int N) {
        //MinPerimeterRectangle
        int minPer=Integer.MAX_VALUE;

        for (int n=1;n<=Math.sqrt(N);n++) {
            if (N%n == 0) {
                int other = N/n;
                minPer = (int)Math.min((2*((long)n+other)),minPer); // cast for possible overflows
            }
        }

        return minPer;
    }

    @DataProvider(name = "test")
    public Object [][] createData1() {
        return new Object [][] {
                new Object [] {30, 22},
        };
    }

    @Test(dataProvider = "test")
    public void verifySolution(int pInput, int expectedFactors) {
        Assert.assertEquals(solution(pInput), expectedFactors);
    }
}
